cloudinit/sources/helpers/azure.py

```diff
@@ -536,42 +669,131 @@ def find_endpoint(fallback_lease_file=None, dhcp245=None):
+        return self._parse_goal_state(
+            unparsed_goal_state_xml, azure_endpoint_client)
+
+    @azure_ds_telemetry_reporter
+    def _get_goal_state_from_azure(self, azure_endpoint_client):
+        """Fetches the GoalState XML from the Azure endpoint and returns
+        the XML as a string.
+
+        @param azure_endpoint_client: instance of AzureEndpointHttpClient
+        @return: GoalState XML string
+        """
+
         LOG.info('Registering with Azure...')
-        attempts = 0
-        while True:
-            try:
-                response = http_client.get(
-                    'http://{0}/machine/?comp=goalstate'.format(self.endpoint))
-            except Exception as e:
-                if attempts < 10:
-                    time.sleep(attempts + 1)
-                else:
-                    report_diagnostic_event(
-                        "failed to register with Azure: %s" % e)
-                    raise
-            else:
-                break
-            attempts += 1
+        url = 'http://{0}/machine/?comp=goalstate'.format(self.endpoint)
+        try:
+            response = azure_endpoint_client.get(url)
+        except Exception as e:
+            msg = 'failed to register with Azure: %s' % e
+            LOG.warning(msg)
+            report_diagnostic_event(msg)
+            raise
         LOG.debug('Successfully fetched GoalState XML.')
-        goal_state = GoalState(response.contents, http_client)
-        report_diagnostic_event("container_id %s" % goal_state.container_id)
+        return response.contents
+
+    @azure_ds_telemetry_reporter
+    def _parse_goal_state(self,
+                          unparsed_goal_state_xml,
+                          azure_endpoint_client):
+        """Parses a GoalState XML string and returns a GoalState object.
+
+        @param unparsed_goal_state_xml: GoalState XML string
+        @param azure_endpoint_client: instance of AzureEndpointHttpClient
+        @return: GoalState object representing the GoalState XML
+        """
+        try:
+            goal_state = GoalState(
+                unparsed_goal_state_xml, azure_endpoint_client)
+        except Exception as e:
+            msg = 'Error processing GoalState XML: %s' % e
+            LOG.warning(msg)
+            report_diagnostic_event(msg)
+            raise
+        msg = ', '.join([
+            'GoalState XML container id: %s' % goal_state.container_id,
+            'GoalState XML instance id: %s' % goal_state.instance_id,
+            'GoalState XML incarnation: %s' % goal_state.incarnation])
+        LOG.debug(msg)
+        report_diagnostic_event(msg)
+        return goal_state
+
+    @azure_ds_telemetry_reporter
+    def _get_user_pubkeys(self, goal_state, pubkey_info):
+        """Gets and filters the VM user's authorized pubkeys.
+
+        cloud-init expects a straightforward array of keys to be dropped
+        into the user's authorized_keys file. Azure control plane exposes
```

Comment:
The "user" here is the "admin user" that is specified when a VM is deployed on Azure. See https://docs.microsoft.com/en-us/cli/azure/vm?view=azure-cli-latest#az-vm-create and https://docs.microsoft.com/en-us/cli/azure/vm#az-vm-create.

I have since updated the docs to reflect this.